### PR TITLE
[ALLUXIO-1910] Refactor Alluxio shell to respect properties passed in command-line

### PR DESCRIPTION
**Pull request by apc999**
https://alluxio.atlassian.net/browse/ALLUXIO-1910

being able to do
```bash
$ bin/alluxio fs -D alluxio.user.file.writetype.default=THROUGH mkdir /foo
```
and foo will show up in under storage

This PR also has some minor formatting change / code clean up for shell commands.

### CODE REVIEW

File: tests/src/test/java/alluxio/shell/command/ChgrprCommandTest.java [hunk 1]
```diff
@@ -1,43 +0,0 @@
-/*
```
Comment (apc999) [inline tests/src/test/java/alluxio/shell/command/ChgrprCommandTest.java]: already covered in `ChgrpCommandTest`

Comment (apc999) [inline tests/src/test/java/alluxio/shell/command/ChgrprCommandTest.java]: apply the below two deleted tests too

File: shell/src/main/java/alluxio/shell/AlluxioShell.java [hunk 2]
```diff
@@ -41,12 +41,9 @@
  * Class for handling command line inputs.
  */
 @NotThreadSafe
-public class AlluxioShell implements Closeable {
+public final class AlluxioShell implements Closeable {
   private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
   private static final Map<String, String[]> CMD_ALIAS = ImmutableMap.<String, String[]>builder()
-      .put("chgrpr", new String[] {"chgrp", "-R"})
```
Comment (apc999) [inline shell/src/main/java/alluxio/shell/AlluxioShell.java]: we never officially support these three commands in releases

File: shell/src/main/java/alluxio/shell/command/ChmodCommand.java [hunk 3]
```diff
@@ -87,7 +87,7 @@ public void run(CommandLine cl) throws IOException {
 
   @Override
   public String getUsage() {
-    return "chmod -R <mode> <path>";
```
Comment (apc999) [inline shell/src/main/java/alluxio/shell/command/ChmodCommand.java]: keep the usage string consistent with other usage strings of shell commands.

File: bin/alluxio [hunk 4]
```diff
@@ -8,8 +8,11 @@ fi
 BIN=$(cd "$( dirname "$0" )"; pwd)
 
 function printUsage {
-  echo "Usage: alluxio COMMAND "
-  echo "where COMMAND is one of:"
+  echo "Usage: alluxio COMMAND [GENERIC_OPTIONS] [COMMAND_OPTIONS]"
+  echo "GENERIC_OPTIONS supports:"
+  echo -e "  -D<property=value>\tUse value for a given property"
```
Comment (ifcharming) [inline bin/alluxio]: nit: add a space after \t ?

File: bin/alluxio [hunk 5]
```diff
@@ -40,6 +43,21 @@ fi
 COMMAND=$1
 shift
 
+ALLUXIO_SHELL_JAVA_OPTS=""
+while getopts ":D:" opt; do
+  case $opt in
```
Comment (ifcharming) [inline bin/alluxio]: nit: ${opt}

File: bin/alluxio [hunk 6]
```diff
@@ -40,6 +43,21 @@ fi
...
+  D)
+    ALLUXIO_SHELL_JAVA_OPTS+=" -D$OPTARG"
```
Comment (ifcharming) [inline bin/alluxio]: nit: ${OPTARG}

File: bin/alluxio [hunk 7]
```diff
@@ -40,6 +43,21 @@ fi
...
+    ;;
+  \?)
+    echo "Invalid option: -$OPTARG"
```
Comment (ifcharming) [inline bin/alluxio]: nit: ${OPTARG}

File: bin/alluxio [hunk 8]
```diff
@@ -40,6 +43,21 @@ fi
...
+    printUsage
+    exit 1
+    ;;
+  esac
+done
+shift $(expr $OPTIND - 1 )
```
Comment (ifcharming) [inline bin/alluxio]: nit: shift $((${OPTIND} - 1)) is preferred.

Per google style guide:
Prefer this:
addition=$((${X} + ${Y}))
substitution="${string/#foo/bar}"

**Comment by ifcharming** [comment]
Thanks @apc999 for the improvement! I left a few minor comments.

File: bin/alluxio (hunk 4 above)
Comment (apc999) [inline bin/alluxio]: done

File: bin/alluxio (hunk 5 above)
Comment (apc999) [inline bin/alluxio]: done

File: bin/alluxio (hunk 6 above)
Comment (apc999) [inline bin/alluxio]: done

File: bin/alluxio (hunk 7 above)
Comment (apc999) [inline bin/alluxio]: done

File: bin/alluxio (hunk 8 above)
Comment (apc999) [inline bin/alluxio]: thanks for the explanation. fixed.

**Comment by apc999** [comment]
@ifcharming your comments addressed. PTAL. thanks

File: bin/alluxio [hunk 9]
```diff
@@ -8,9 +8,12 @@ fi
 BIN=$(cd "$( dirname "$0" )"; pwd)
 
 function printUsage {
-  echo "Usage: alluxio COMMAND "
-  echo "where COMMAND is one of:"
-  echo -e "  format [-s]   \t Format Alluxio (if -s specified, only format if underfs is local and doesn't already exist)"
+  echo "Usage: alluxio COMMAND [GENERIC_OPTIONS] [COMMAND_OPTIONS]"
```
Comment (aaudiber) [inline bin/alluxio]: Could we combine this into just [OPTIONS]? I've never heard of a tool with two types of options which need to be in a specific order

File: bin/alluxio [hunk 10]
```diff
@@ -8,9 +8,12 @@ fi
...
+  echo "GENERIC_OPTIONS supports:"
+  echo -e "  -D<property=value>\t Use value for a given Alluxio property"
```
Comment (aaudiber) [inline bin/alluxio]: Use a value for a given Alluxio property

**Comment by aaudiber** [comment]
Thanks for improving the configuration story for the shell! I left a couple comments

File: shell/src/main/java/alluxio/shell/command/FreeCommand.java [hunk 11]
```diff
@@ -58,11 +58,11 @@ void runCommand(AlluxioURI path, CommandLine cl) throws IOException {
...
   @Override
   public String getUsage() {
-    return "free <file path|folder path>";
+    return "free <path>";
   }
 
   @Override
   public String getDescription() {
-    return "Removes the file or directory(recursively) from Alluxio memory space.";
+    return "Frees the space of a file or directory(recursively) from Alluxio memory space.";
```
Comment (jsimsa) [inline shell/src/main/java/alluxio/shell/command/FreeCommand.java]: add space between `directory` and `(recursively)`

Comment (jsimsa) [inline shell/src/main/java/alluxio/shell/command/FreeCommand.java]: Actually I would reword this to: `Frees the space occupied by a file or a directory in Alluxio.`

File: shell/src/main/java/alluxio/shell/command/ChownCommand.java [hunk 12]
```diff
@@ -85,7 +85,7 @@ public void run(CommandLine cl) throws IOException {
 
   @Override
   public String getUsage() {
-    return "chown -R <owner> <path>";
+    return "chown [-R] <owner> <path>";
```
Comment (jsimsa) [inline shell/src/main/java/alluxio/shell/command/ChownCommand.java]: do we need similar change for `chgrp`?

File: bin/alluxio (hunk 9 above)
Comment (jsimsa) [inline bin/alluxio]: can we make the usage a variable (similar to how it is done in alluxio-start.sh

Comment (jsimsa) [inline bin/alluxio]: as for @aaudiber's comment ... I think the following pattern is common:

`<command> <command_options> <subcommand> <subcommand_options> ...`

so to be consistent with that, the invocation we support should look like:

`alluxio [ALLUXIO_OPTIONS] COMMAND [COMMAND_OPTIONS]`

having said that, since not all the commands actually use -D (e.g. bootstrap-conf or killAll or copyDir), I would not call these out specifically ... I think you do that because of the limitations of parsing options in shell, correct?

this would be much easier in Go =_=

**Comment by jsimsa** [comment]
@apc999 thanks, did a pass

File: shell/src/main/java/alluxio/shell/command/ChownCommand.java (hunk 12 above)
Comment (apc999) [inline shell/src/main/java/alluxio/shell/command/ChownCommand.java]: chgrp is already having such a usage message ('[R]')

Comment (jsimsa) [inline shell/src/main/java/alluxio/shell/command/ChownCommand.java]: ack

File: bin/alluxio (hunk 10 above)
Comment (apc999) [inline bin/alluxio]: done

**Comment by apc999** [comment]
@jsimsa  PTAL. thanks

File: bin/alluxio [hunk 13]
```diff
@@ -278,64 +217,42 @@ case ${COMMAND} in
...
 "formatWorker")
   echo "Formatting Alluxio Worker @ `hostname -f`"
-  CLASS=alluxio.Format
-  PARAMETER=worker
+  CLASS="alluxio.cli.Format"
+  PARAMETER="worker"
 ;;
 "fs")
-  CLASS=alluxio.shell.AlluxioShell
+  CLASS="alluxio.shell.AlluxioShell"
```
Comment (jsimsa) [inline bin/alluxio]: alluxio.shell.cli.AlluxioShell?

File: bin/alluxio [hunk 14]
```diff
@@ -278,64 +217,42 @@ case ${COMMAND} in
...
 ;;
 "loadufs")
-  CLASS=alluxio.client.UfsUtils
+  CLASS="alluxio.client.UfsUtils"
```
Comment (jsimsa) [inline bin/alluxio]: alluxio.client.cli.UfsUtils?

File: bin/alluxio [hunk 15]
```diff
@@ -370,15 +287,30 @@ case ${COMMAND} in
...
 ;;
 *)
+  echo "Unsupported command ${COMMAND}"
   printUsage
   exit 1
 ;;
 esac
 
-${JAVA} -cp ${CLASSPATH} ${ALLUXIO_USER_JAVA_OPTS} ${CLASS} ${PARAMETER} $@
+while getopts ":D:" opt; do
```
Comment (jsimsa) [inline bin/alluxio]: factor this into a function and only call it for commands that will end up calling `java`

Comment (jsimsa) [inline bin/alluxio]: for readability sake, factor the the remaining logic into a function (e.g. `runJavaClass`) and invoke it from the relevant commands

File: bin/alluxio [hunk 16]
```diff
@@ -247,6 +185,7 @@ function copyDir {
 }
 
 PARAMETER=""
+ALLUXIO_SHELL_JAVA_OPTS=""
 
 case ${COMMAND} in
```
Comment (jsimsa) [inline bin/alluxio]: for readability sake wrap this switch command in a main function that gets invoked from the bottom of the shell script (https://google.github.io/styleguide/shell.xml#main)

Comment (jsimsa) [inline bin/alluxio]: basically the only non-function control flow in this script should be:

`main "$@"`

File: bin/alluxio [hunk 17]
```diff
@@ -8,9 +8,10 @@ fi
 BIN=$(cd "$( dirname "$0" )"; pwd)
 
 function printUsage {
-  echo "Usage: alluxio COMMAND "
-  echo "where COMMAND is one of:"
-  echo -e "  format [-s]   \t Format Alluxio (if -s specified, only format if underfs is local and doesn't already exist)"
+  echo "Usage: alluxio COMMAND [GENERIC_COMMAND_OPTIONS] [COMMAND_ARGS]"
```
Comment (jsimsa) [inline bin/alluxio]: the -D options is not meaningful for all commands ... so I think the current usage is misleading ... it might make more sense to group the commands in two groups (one that ends up starting a JVM and another that does not) and use `JAVA_OPTIONS` instead of GENERIC_COMMAND_OPTIONS ... also, it might make sense to do the following:

`alluxio COMMAND [JAVA_OPTIONS --] [COMMAND_ARGS]` so that we can simply look for `--` to decide which arguments are JVM options and which arguments should be passed to the Java program

File: conf/log4j.properties [hunk 18]
```diff
@@ -1,5 +1,4 @@
 # May get overridden by System Property
-alluxio.logger.type=Console
```
Comment (jsimsa) [inline conf/log4j.properties]: OOC, what is the effect of this change?

File: core/common/src/main/java/alluxio/AbstractClient.java [hunk 19]
```diff
@@ -11,6 +11,7 @@
 
 package alluxio;
 
+import alluxio.cli.Version;
```
Comment (jsimsa) [inline core/common/src/main/java/alluxio/AbstractClient.java]: remove?

File: core/server/src/main/java/alluxio/cli/Format.java [hunk 20]
```diff
@@ -29,8 +32,8 @@
 @ThreadSafe
 public final class Format {
   private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
-  private static final String USAGE = String.format("java -cp %s alluxio.Format <MASTER/WORKER>",
-      Version.ALLUXIO_JAR);
+  private static final String USAGE = String.format("java -cp %s %s <MASTER/WORKER>",
```
Comment (jsimsa) [inline core/server/src/main/java/alluxio/cli/Format.java]: might make sense to use the same pattern as AlluxioMaster and AlluxioWorker

File: core/server/src/main/java/alluxio/cli/Format.java [hunk 21]
```diff
@@ -51,7 +54,7 @@ private static boolean formatFolder(String name, String folder, Configuration co
   }
 
   /**
-   * Formats the Alluxio file system via {@code java -cp %s alluxio.Format <MASTER/WORKER>}.
+   * Formats the Alluxio file system via {@code java -cp %s alluxio.cli.Format <MASTER/WORKER>}.
```
Comment (jsimsa) [inline core/server/src/main/java/alluxio/cli/Format.java]: if we inline the USAGE message here, we can simplify this to `Format the Alluxio system.`

File: examples/src/main/java/alluxio/cli/CliUtils.java [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+
+import java.util.concurrent.Callable;
+
+/**
+ * Utilities to run the examples.
+ */
+public final class CliUtils {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  private CliUtils() {}
```
Comment (jsimsa) [inline examples/src/main/java/alluxio/cli/CliUtils.java]: // prevent instantiation

File: examples/src/main/java/alluxio/cli/JournalCrashTest.java [hunk 23]
```diff
@@ -234,7 +233,7 @@ private static void killMaster() {
    *
    * Usage:
    * {@code java -cp
-   * alluxio-<ALLUXIO-VERSION>-jar-with-dependencies.jar alluxio.examples.JournalCrashTest}
+   * alluxio-<ALLUXIO-VERSION>-jar-with-dependencies.jar alluxio.cli.JournalCrashTest}
```
Comment (jsimsa) [inline examples/src/main/java/alluxio/cli/JournalCrashTest.java]: instead of adding a comment here, print a usage message in the main function when it fails (similar to how AlluxioMaster or AlluxioWorker do this)

File: examples/src/main/java/alluxio/cli/RunTests.java [hunk 24]
```diff
@@ -0,0 +1,154 @@
...
+    System.out.println("(1) To run a predefined set of read/write tests:");
+    System.out.println(String
+        .format("java -cp %s %s <master address>", Version.ALLUXIO_JAR, RunTests.class.getName()));
+    System.out.println("(2) To run a single test:");
+    System.out.println(String
+        .format("java -cp %s %s <master address> <%s> <%s> <%s>", Version.ALLUXIO_JAR,
+            RunTests.class.getName(), OperationType.class.getSimpleName(),
+            ReadType.class.getSimpleName(), WriteType.class.getSimpleName()));
+    System.out.println(String
+        .format("\t %s: one of %s", OperationType.class.getSimpleName(),
```
Comment (jsimsa) [inline examples/src/main/java/alluxio/cli/RunTests.java]: nit replace `\t %s ...` with `\t <%s> ...`

File: examples/src/main/java/alluxio/cli/RunTests.java [hunk 25]
```diff
@@ -0,0 +1,154 @@
...
+            Arrays.toString(OperationType.values())));
+    System.out.println(
+        String.format("\t %s: one of %s", ReadType.class.getSimpleName(), READ_TYPES));
```
Comment (jsimsa) [inline examples/src/main/java/alluxio/cli/RunTests.java]: ditto

File: examples/src/main/java/alluxio/cli/RunTests.java [hunk 26]
```diff
@@ -0,0 +1,154 @@
...
+    System.out.println(String
+        .format("\t %s: one of %s", WriteType.class.getSimpleName(), WRITE_TYPES));
```
Comment (jsimsa) [inline examples/src/main/java/alluxio/cli/RunTests.java]: ditto

File: examples/src/main/java/alluxio/cli/RunTests.java [hunk 27]
```diff
@@ -0,0 +1,154 @@
...
+  /** Write types to test. */
+  private static final List<WriteType> WRITE_TYPES = Arrays
+      .asList(WriteType.MUST_CACHE, WriteType.CACHE_THROUGH, WriteType.THROUGH,
+          WriteType.ASYNC_THROUGH);
+
+  private static void usage() {
+    System.out.println("Usage:");
+    System.out.println("(1) To run a predefined set of read/write tests:");
+    System.out.println(String
+        .format("java -cp %s %s <master address>", Version.ALLUXIO_JAR, RunTests.class.getName()));
```
Comment (jsimsa) [inline examples/src/main/java/alluxio/cli/RunTests.java]: ooc, what is the difference between `getName()` and `getCanonicalName()` (which is used elsewhere to get the fully qualified class name

File: examples/src/main/java/alluxio/cli/RunTests.java [hunk 28]
```diff
@@ -0,0 +1,154 @@
...
+   * @param writeType write type
+   * @return 0 on success, 1 on failure
+   */
+  private static int runTest(AlluxioURI masterLocation, OperationType opType, ReadType readType,
+      WriteType writeType) {
+    AlluxioURI filePath =
+        new AlluxioURI(String.format("%s/%s_%s_%s", TEST_PATH, opType, readType, writeType));
+
+    boolean result = true;
+    if (opType == OperationType.Basic) {
```
Comment (jsimsa) [inline examples/src/main/java/alluxio/cli/RunTests.java]: nit: consider using switch for the enum

File: examples/src/main/java/alluxio/examples/keyvalue/KeyValueStoreOperations.java [hunk 29]
```diff
@@ -14,14 +14,14 @@
...
-import alluxio.Version;
+import alluxio.cli.Version;
 import alluxio.client.ClientContext;
 import alluxio.client.keyvalue.KeyValueIterator;
 import alluxio.client.keyvalue.KeyValuePair;
 import alluxio.client.keyvalue.KeyValueStoreReader;
 import alluxio.client.keyvalue.KeyValueStoreWriter;
 import alluxio.client.keyvalue.KeyValueSystem;
-import alluxio.examples.Utils;
+import alluxio.cli.CliUtils;
```
Comment (jsimsa) [inline examples/src/main/java/alluxio/examples/keyvalue/KeyValueStoreOperations.java]: sort imports

File: examples/src/main/java/alluxio/examples/keyvalue/KeyValueStoreOperations.java [hunk 30]
```diff
@@ -132,6 +132,7 @@ public static void main(String[] args) throws Exception {
       System.exit(-1);
     }
 
-    Utils.runExample(new KeyValueStoreOperations(new AlluxioURI(args[0])));
+    boolean result = CliUtils.runExample(new KeyValueStoreOperations(new AlluxioURI(args[0])));
```
Comment (jsimsa) [inline examples/src/main/java/alluxio/examples/keyvalue/KeyValueStoreOperations.java]: the "run and exit" pattern shows up repeatedly in the code base and it might make sense to add a utility function for it to `CliUtils`

**Comment by jsimsa** [comment]
@apc999 thank you, I think this is a great improvement ... did a pass and left a number of comments

File: bin/alluxio (hunk 17 above)
Comment (apc999) [inline bin/alluxio]: as discussed offline, we will have a followup PR to fix this by having more detailed help message. 
a TODO added

File: bin/alluxio (hunk 16 above)
Comment (apc999) [inline bin/alluxio]: fixed

File: bin/alluxio (hunk 13 above)
Comment (apc999) [inline bin/alluxio]: moved to `alluxio.cli`

File: bin/alluxio (hunk 14 above)
Comment (apc999) [inline bin/alluxio]: moved to `alluxio.cli` in shell package from core-client package

File: bin/alluxio (hunk 16 above)
Comment (apc999) [inline bin/alluxio]: fixed

File: bin/alluxio (hunk 15 above)
Comment (apc999) [inline bin/alluxio]: fixed

File: core/server/src/main/java/alluxio/cli/Format.java (hunk 20 above)
Comment (apc999) [inline core/server/src/main/java/alluxio/cli/Format.java]: we do it a separate SMALLFIX PR

File: core/server/src/main/java/alluxio/cli/Format.java (hunk 21 above)
Comment (apc999) [inline core/server/src/main/java/alluxio/cli/Format.java]: removed

File: examples/src/main/java/alluxio/cli/CliUtils.java (hunk 22 above)
Comment (apc999) [inline examples/src/main/java/alluxio/cli/CliUtils.java]: added

File: examples/src/main/java/alluxio/cli/JournalCrashTest.java (hunk 23 above)
Comment (apc999) [inline examples/src/main/java/alluxio/cli/JournalCrashTest.java]: done

File: examples/src/main/java/alluxio/cli/RunTests.java (hunk 27 above)
Comment (apc999) [inline examples/src/main/java/alluxio/cli/RunTests.java]: `getCanonicalName()` should be used, I think, for this particular usage, they are the same.
http://stackoverflow.com/questions/15202997/what-is-the-difference-between-canonical-name-simple-name-and-class-name-in-jav

File: examples/src/main/java/alluxio/cli/RunTests.java (hunk 24 above)
Comment (apc999) [inline examples/src/main/java/alluxio/cli/RunTests.java]: done

File: examples/src/main/java/alluxio/cli/RunTests.java (hunk 25 above)
Comment (apc999) [inline examples/src/main/java/alluxio/cli/RunTests.java]: done

File: examples/src/main/java/alluxio/cli/RunTests.java (hunk 26 above)
Comment (apc999) [inline examples/src/main/java/alluxio/cli/RunTests.java]: done

File: examples/src/main/java/alluxio/cli/RunTests.java (hunk 28 above)
Comment (apc999) [inline examples/src/main/java/alluxio/cli/RunTests.java]: done

File: examples/src/main/java/alluxio/examples/keyvalue/KeyValueStoreOperations.java (hunk 29 above)
Comment (apc999) [inline examples/src/main/java/alluxio/examples/keyvalue/KeyValueStoreOperations.java]: done

File: examples/src/main/java/alluxio/examples/keyvalue/KeyValueStoreOperations.java (hunk 30 above)
Comment (apc999) [inline examples/src/main/java/alluxio/examples/keyvalue/KeyValueStoreOperations.java]: A TODO left

**Comment by apc999** [comment]
@jsimsa @aaudiber @ifcharming , I fixed most of the comments. PTAL. 
thanks

**Comment by apc999** [comment]
Jenkins, test this please

File: bin/alluxio [hunk 31]
```diff
@@ -246,139 +160,172 @@ function copyDir {
...
+
+  if [[ $# == 0 ]]; then
+    printUsage
+    exit 1
+  fi
+
+  COMMAND=$1
+  shift
+
+  # bootstrap creates alluxio-env.sh
```
Comment (jsimsa) [inline bin/alluxio]: add a comment that mentions that the needs to happens before source alluxio-config.sh

**Comment by jsimsa** [comment]
@apc999 just one minor comment, otherwise LGTM

**Comment by apc999** [comment]
@jsimsa comment addressed.

**Comment by jsimsa** [comment]
LGTM

**Comment by aaudiber** [comment]
LGTM
